src/v/cloud_storage/inventory/report_parser.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "bytes/details/io_allocation_size.h"
+
+#include <seastar/core/iostream.hh>
+
+#include <zlib.h>
+
+namespace cloud_storage::inventory {
+
+using gzip_compressed_t = ss::bool_class<struct gzip_compressed_tag>;
+
+class report_decompression_error : public std::runtime_error {
+public:
+    using runtime_error::runtime_error;
+};
+
+/// Parses newline delimited inventory report, supplied as an input stream. The
+/// report may optionally be GZip compressed. The rows returned by the parser
+/// have newlines stripped off, but no other modifications are performed. In
+/// particular, whitespaces and carriage returns are still present in the
+/// returned row. For CSV files, parsing of the row is not performed and must be
+/// done by the caller.
+class report_parser {
+public:
+    report_parser(const report_parser&) = delete;
+    report_parser& operator=(const report_parser&) = delete;
+
+    report_parser(report_parser&&) = delete;
+    report_parser& operator=(report_parser&&) = delete;
+
+    // The chunk size supplied here controls the following behavior:
+    // 1. The size of data read from the input stream per read_up_to invocation
+    // 2. The max buffer size allocated during chunk decompression if the report
+    // is compressed
+    explicit report_parser(
+      ss::input_stream<char> stream,
+      gzip_compressed_t is_compressed = gzip_compressed_t::no,
+      size_t chunk_size = details::io_allocation_size::max_chunk_size);
+
+    // Returns the next line from the input stream, or nullopt if the stream is
+    // fully consumed.
+    ss::future<std::optional<ss::sstring>> next_line();
+    bool eof() const { return _eof; }
+
+    // High level API for consuming a report. The consumer is supplied with each
+    // row in the stream until EOF. The consumer must return a future;
+    template<typename Consumer>
+    ss::future<> consume(Consumer c) {
+        for (auto row = co_await next_line(); row.has_value();
+             row = co_await next_line()) {
+            co_await c(row.value());
+        }
+    }
+
+    ~report_parser();
```

Comment:
do we need a gate for async operations in this class?

Reply:
added gate